File: leetcode/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
import unittest
from typing import List

class Solution:
    def _binary_search_left(self, nums: List[int], target: int, left: int, right: int):
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > target:
                right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid
        return left if nums[left] == target else -1
        
    def _binary_search_right(self, nums: List[int], target: int, left: int, right: int):
        while left < right:
            mid = (left + right + 1) // 2
            if nums[mid] > target:
                right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                left = mid
        return left if nums[left] == target else -1

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        n, left, right = len(nums), -1, -1
        if n != 0:
            left = self._binary_search_left(nums, target, 0, n - 1)
            if left != -1:
                right = self._binary_search_right(nums, target, left, n - 1)
        return [left, right]

class TestSearchRange(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().searchRange([5,7,7,8,8,10], 8), [3,4])    

    def test_2(self):
        self.assertEqual(Solution().searchRange([5,7,7,8,8,10], 6), [-1, -1])    

    def test_3(self):
        self.assertEqual(Solution().searchRange([], 0), [-1, -1])    

class TestBinarySearchLeft(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution()._binary_search_left([2,5,5,5,5,8], 5, 0, 5), 1)       

    def test_2(self):
        self.assertEqual(Solution()._binary_search_left([5,5,5,5,5,8], 5, 0, 5), 0)       

    def test_3(self):
        self.assertEqual(Solution()._binary_search_left([5,5,5,5,5,8], 6, 0, 5), -1)       

class TestBinarySearchRight(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution()._binary_search_right([2,5,5,5,5,8], 5, 0, 5), 4)       

    def test_2(self):
        self.assertEqual(Solution()._binary_search_right([4, 5,5,5,5,5], 5, 0, 5), 5)       

    def test_3(self):
        self.assertEqual(Solution()._binary_search_right([5,5,5,5,5,8], 6, 0, 5), -1)       

File: leetcode/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.rs
struct Solution {}

impl Solution {
    fn _binary_search_left(nums: &Vec<i32>, target: i32, mut left: usize, mut right: usize) -> Option<usize> {
        while left < right {
            let mid: usize = (left + right) / 2;
            if nums[mid] > target {
                right = mid;
            } else if nums[mid] < target {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if nums[left] == target { Some(left) } else { None }
    }

    fn _binary_search_right(nums: &Vec<i32>, target: i32, mut left: usize, mut right: usize) -> Option<usize> {
        while left < right {
            let mid: usize = (left + right + 1) / 2;
            if nums[mid] > target {
                right = mid - 1;
            } else if nums[mid] < target {
                left = mid + 1;
            } else {
                left = mid;
            }
        }
        if nums[left] == target { Some(left) } else { None }
    }

    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let not_found = vec![-1, -1];
        match nums.len() {
            0 => not_found,
            n => match Self::_binary_search_left(&nums, target, 0, n - 1) {
                None => not_found,
                Some(left) => match Self::_binary_search_right(&nums, target, left, n - 1) {
                    None => not_found,
                    Some(right) => vec![left as i32, right as i32],
                },
            },
        }
    }
}

#[cfg(test)]
mod search_range_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::search_range(vec![5,7,7,8,8,10], 8), vec![3, 4]);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::search_range(vec![5,7,7,8,8,10], 6), vec![-1, -1]);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::search_range(vec![], 8), vec![-1, -1]);
    }

    #[test]
    fn test_4() {
        assert_eq!(Solution::search_range(vec![2, 2], 1), vec![-1, -1]);
    }
}
